packages/api/src/providers/ActivityKeyer/ActivityKeyerComposer.tsx

```diff
@@ -43,53 +44,36 @@ const ActivityKeyerComposer = ({ children }: Readonly<{ children?: ReactNode | u
-    const nextActivityToKeyMap: ActivityToKeyMap = new Map();
-    const nextClientActivityIdToKeyMap: ClientActivityIdToKeyMap = new Map();
-    const nextKeyToActivitiesMap: KeyToActivitiesMap = new Map();
-
-    activities.forEach(activity => {
+  // Process new activities (if any)
+  if (activities.length > lastProcessedIndexRef.current) {
+    for (let i = lastProcessedIndexRef.current; i < activities.length; i++) {
+      // eslint-disable-next-line security/detect-object-injection
+      const activity = activities[i];
       const activityId = getActivityId(activity);
+      if (!activityId) {
+        break;
+      }
       const clientActivityId = getClientActivityId(activity);
       const typingActivityId = getActivityLivestreamingMetadata(activity)?.sessionId;
 
       const key =
-        (clientActivityId &&
-          (clientActivityIdToKeyMap.get(clientActivityId) || nextClientActivityIdToKeyMap.get(clientActivityId))) ||
-        (typingActivityId &&
-          (activityIdToKeyMap.get(typingActivityId) || nextActivityIdToKeyMap.get(typingActivityId))) ||
-        (activityId && (activityIdToKeyMap.get(activityId) || nextActivityIdToKeyMap.get(activityId))) ||
-        activityToKeyMap.get(activity) ||
-        nextActivityToKeyMap.get(activity) ||
+        (clientActivityId && clientActivityIdToKeyMapRef.current.get(clientActivityId)) ||
+        (typingActivityId && activityIdToKeyMapRef.current.get(typingActivityId)) ||
+        (activityId && activityIdToKeyMapRef.current.get(activityId)) ||
+        activityToKeyMapRef.current.get(activity) ||
         uniqueId();
 
-      activityId && nextActivityIdToKeyMap.set(activityId, key);
-      clientActivityId && nextClientActivityIdToKeyMap.set(clientActivityId, key);
-      nextActivityToKeyMap.set(activity, key);
-      nextActivityKeys.add(key);
-
-      const activities = nextKeyToActivitiesMap.has(key) ? [...nextKeyToActivitiesMap.get(key)] : [];
-
-      activities.push(activity);
-      nextKeyToActivitiesMap.set(key, Object.freeze(activities));
-    });
-
-    activityIdToKeyMapRef.current = Object.freeze(nextActivityIdToKeyMap);
-    activityToKeyMapRef.current = Object.freeze(nextActivityToKeyMap);
-    clientActivityIdToKeyMapRef.current = Object.freeze(nextClientActivityIdToKeyMap);
-    keyToActivitiesMapRef.current = Object.freeze(nextKeyToActivitiesMap);
-
-    // `nextActivityKeys` could potentially same as `prevActivityKeys` despite reference differences, we should memoize it.
-    return Object.freeze([Object.freeze([...nextActivityKeys.values()])]) as readonly [readonly string[]];
-  }, [activities, activityIdToKeyMapRef, activityToKeyMapRef, clientActivityIdToKeyMapRef, keyToActivitiesMapRef]);
+      activityId && activityIdToKeyMapRef.current.set(activityId, key);
+      clientActivityId && clientActivityIdToKeyMapRef.current.set(clientActivityId, key);
 
+      activityToKeyMapRef.current.set(activity, key);
```

Comment:
I am curious why this would work.

`activityToKeyMapRef` is frozen and setting to it should throw (in strict mode) or no-op (in non-strict mode).